allennlp/semparse/type_declarations/wikitables_variable_free.py

```diff
@@ -0,0 +1,126 @@
+"""
+Defines all types in the variable-free language for the WikiTablesQuestions dataset defined in the
+following paper by Liang et al. (2018)
+Memory Augmented Policy Optimization for Program Synthesis with Generalization.
+"""
+
+
+from allennlp.semparse.type_declarations.type_declaration import NamedBasicType, ComplexType
+
+
+# Basic types
+# Note that there aren't cell types and part types in this gramamr. They are all of type string. The
+# constants will still come from the `TableQuestionKnowledgeGraph`, and are identified as parts and
+# cells, and the executor can process them accordingly, but the grammar does not differentiate
+# between them. Also note that columns are basic types in this grammar.
+ROW_TYPE = NamedBasicType("ROW")
+# "LCOLUMN" to ensure the signature will be "l", to avoid confusion with cell. I decided to omit the
+# cell type eventually, and the decision to call this "LCOLUMN" is not relevant any more. But I'm
+# not changing it for now, because I might end up adding the cell type back later.
+COLUMN_TYPE = NamedBasicType("LCOLUMN")
+NUMBER_TYPE = NamedBasicType("NUMBER")
+DATE_TYPE = NamedBasicType("DATE")
+STRING_TYPE = NamedBasicType("STRING")
+
+BASIC_TYPES = {ROW_TYPE, COLUMN_TYPE, NUMBER_TYPE, DATE_TYPE, STRING_TYPE}
+STARTING_TYPES = {NUMBER_TYPE, DATE_TYPE, STRING_TYPE}
+
+# Complex types
+# Type for selecting the value in a column in a set of rows. "select" and "mode" functions.
+SELECT_TYPE = ComplexType(ROW_TYPE, ComplexType(COLUMN_TYPE, STRING_TYPE))
+
+# Type for filtering rows given a column. "argmax", "argmin" and "same_as" (select all rows with the
+# same value under the given column as the given row does under the given column)
+ROW_FILTER_WITH_COLUMN = ComplexType(ROW_TYPE, ComplexType(COLUMN_TYPE, ROW_TYPE))
+
+# "filter_number_greater", "filter_number_equals" etc.
+ROW_FILTER_WITH_COLUMN_AND_NUMBER = ComplexType(ROW_TYPE,
+                                                ComplexType(NUMBER_TYPE,
+                                                            ComplexType(COLUMN_TYPE, ROW_TYPE)))
+
+# "filter_date_greater", "filter_date_equals" etc.
+ROW_FILTER_WITH_COLUMN_AND_DATE = ComplexType(ROW_TYPE,
+                                              ComplexType(DATE_TYPE,
+                                                          ComplexType(COLUMN_TYPE, ROW_TYPE)))
+
+# "filter_in" and "filter_not_in"
+ROW_FILTER_WITH_COLUMN_AND_STRING = ComplexType(ROW_TYPE,
+                                                ComplexType(STRING_TYPE,
+                                                            ComplexType(COLUMN_TYPE, ROW_TYPE)))
+
+ROW_FILTER = ComplexType(ROW_TYPE, ROW_TYPE)  # first, last, previous, next etc.
+
+COUNT_TYPE = ComplexType(ROW_TYPE, NUMBER_TYPE)
```

Comment:
You can only count rows?

Reply:
Yes. See Appendix A in the MAPO paper.

Reply:
Or see a nice comment here explaining that...? =)

Reply:
But, I did just look at appendix A in the paper, and it makes it clear that there are different kinds of columns.  We should have those type distinctions, right?

Reply:
Sure, I'll add more comments. They make a distinction between string, number, and date columns. I was going to have the executor handle this distinction. That is, for example, if you pass a column without numbers to the `max` function, return null (I think that's what LambdaDCS does if you apply a `cell.cell.number` on a string value). I'm not sure how this can be done with type declaration. Any ideas?

Reply:
Added a comment.